File: c-interface/include/src/IndexBuffer.cpp
#include "Astral.Canvas/Graphics/IndexBuffer.h"
#include "Graphics/IndexBuffer.hpp"
#include "allocators.hpp"

exportC AstralCanvas_IndexBufferSize AstralCanvasIndexBuffer_GetSize(AstralCanvasIndexBuffer ptr)
{
    return (AstralCanvas_IndexBufferSize)((AstralCanvas::IndexBuffer *)ptr)->indexElementSize;
}
exportC usize AstralCanvasIndexBuffer_GetCount(AstralCanvasIndexBuffer ptr)
{
    return ((AstralCanvas::IndexBuffer *)ptr)->indexCount;
}
exportC AstralCanvasIndexBuffer AstralCanvasIndexBuffer_Create(AstralCanvas_IndexBufferSize indexSize, usize indexCount)
{
    AstralCanvas::IndexBuffer *result = (AstralCanvas::IndexBuffer *)GetDefaultAllocator()->Allocate(sizeof(AstralCanvas::IndexBuffer));
    *result = AstralCanvas::IndexBuffer((AstralCanvas::IndexBufferSize)indexSize, indexCount);
    return result;
}
exportC void AstralCanvasIndexBuffer_Deinit(AstralCanvasIndexBuffer ptr)
{
    ((AstralCanvas::IndexBuffer *)ptr)->deinit();
    GetDefaultAllocator()->Free(ptr);
}
exportC void AstralCanvasIndexBuffer_SetData(AstralCanvasIndexBuffer ptr, u8* bytes, usize sizeOfBytes)
{
    ((AstralCanvas::IndexBuffer *)ptr)->SetData(bytes, sizeOfBytes);
}

File: bindings/csharp/IndexBuffer.cs
using System;

namespace Astral.Canvas
{
    public class IndexBuffer : IDisposable
    {
        public IntPtr handle;
        public IndexBufferSize indexSize;
        public uint indexCount;

        public IndexBuffer(IndexBufferSize indexSize, uint indexCount)
        {
            this.indexSize = indexSize;
            this.indexCount= indexCount;
            this.handle = AstralCanvas.IndexBuffer_Create(indexSize, (UIntPtr)indexCount);
        }
        public unsafe void SetData<T>(T[] indices) where T : unmanaged
        {
            fixed (T* ptr = indices)
            {
                AstralCanvas.IndexBuffer_SetData(handle, (IntPtr)ptr, (UIntPtr)(sizeof(T) * indices.Length));
            }
        }

        public void Dispose()
        {
            if (handle !=  IntPtr.Zero)
            {
                AstralCanvas.IndexBuffer_Deinit(handle);
            }
            handle = IntPtr.Zero;
        }
    }
}
